src/java.base/share/classes/jdk/internal/classfile/instruction/LoadInstruction.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package jdk.internal.classfile.instruction;
+
+import jdk.internal.classfile.CodeElement;
+import jdk.internal.classfile.CodeModel;
+import jdk.internal.classfile.Instruction;
+import jdk.internal.classfile.Opcode;
+import jdk.internal.classfile.TypeKind;
+import jdk.internal.classfile.impl.AbstractInstruction;
+import jdk.internal.classfile.impl.BytecodeHelpers;
+import jdk.internal.classfile.impl.Util;
+
+/**
+ * Models a local variable load instruction in the {@code code} array of a
+ * {@code Code} attribute.  Corresponding opcodes will have a {@code kind} of
+ * {@link Opcode.Kind#LOAD}.  Delivered as a {@link CodeElement} when
+ * traversing the elements of a {@link CodeModel}.
+ */
+public sealed interface LoadInstruction extends Instruction
+        permits AbstractInstruction.BoundLoadInstruction,
+                AbstractInstruction.UnboundLoadInstruction {
+    int slot();
+
+    TypeKind typeKind();
+
+    /**
+     * {@return a local variable load instruction}
+     *
+     * @param kind the type of the value to be loaded
+     * @param slot the local varaible slot to load from
+     */
+    static LoadInstruction of(TypeKind kind, int slot) {
+        return of(BytecodeHelpers.loadOpcode(kind, slot), slot);
+    }
+
+    /**
+     * {@return a local variable load instruction}
+     *
+     * @param op the opcode for the specific type of load instruction,
+     *           which must be of kind {@link Opcode.Kind#LOAD}
+     * @param slot the local varaible slot to load from
+     */
+    static LoadInstruction of(Opcode op, int slot) {
```

Comment:
Should you validate that `slot` is compatible with `op`? (same for the StoreInstruction?)

Reply:
There are three aspects of this kind of validation:
1. similar to `arrayType` there was an architectural decision to less validate so perform faster
2. static instruction factory does not have the information necessary to validate local variable slot, it could be validated in the `CodeBuilder`
3. incompatible slot would most probably prevent stack map generation and such code will fail in the later phase or at last when verification is explicitly called on the generated code